### PR TITLE
TC-1672 Dashboard - User Preferences back end

### PR DESCRIPTION
**Pull request by bxf12315**

### CODE REVIEW

**Comment by gildub** [comment]
@bxf12315, could you please provide a description for the purpose of this PR ?

**Comment by carlosthe19916** [comment]
@bxf12315 Thanks for creating this PR.
I couldn't find endpoints I could test and verify. If it is an error on my side could you please point me to which are the rest endpoints to be used?

If the endpoints are still to be done here are some ideas:
- We can have an endpoint `GET /user-preferences/selected-sboms` that return an array of 0 to 4 elements. e.g.:
```json
[
{
"sbom_id": "abcd",
"sbom_name": "xyz"
},
{
"sbom_id": "abcd",
"sbom_name": "xyz"
},
{
"sbom_id": "abcd",
"sbom_name": "xyz"
},
{
"sbom_id": "abcd",
"sbom_name": "xyz"
}
]
```

- Each time the UI will need to update those 4 selected SBOM the UI can send a PUT request with the new 4 selected sboms. e.g. `PUT /user-preferences/selected-sboms` with a body like

```json
[
{
"sbom_id": "abcd",
"sbom_name": "xyz"
},
{
"sbom_id": "abcd",
"sbom_name": "xyz"
},
{
"sbom_id": "abcd",
"sbom_name": "xyz"
},
{
"sbom_id": "abcd",
"sbom_name": "xyz"
}
]
```

The token always arrives to the backend and the backend could decode the token and extract the username or userid, or any other field that we might find appropriate.

It would be nice to have the whole thing done in this PR so we avoid having multiple PRs.

What do you think?

**Comment by bxf12315** [comment]
> @bxf12315 Thanks for creating this PR. I couldn't find endpoints I could test and verify. If it is an error on my side could you please point me to which are the rest endpoints to be used?
> 
> If the endpoints are still to be done here are some ideas:
> 
> * We can have an endpoint `GET /user-preferences/selected-sboms` that return an array of 0 to 4 elements. e.g.:
> 
> ```json
> [
> {
> "sbom_id": "abcd",
> "sbom_name": "xyz"
> },
> {
> "sbom_id": "abcd",
> "sbom_name": "xyz"
> },
> {
> "sbom_id": "abcd",
> "sbom_name": "xyz"
> },
> {
> "sbom_id": "abcd",
> "sbom_name": "xyz"
> }
> ]
> ```
> 
> * Each time the UI will need to update those 4 selected SBOM the UI can send a PUT request with the new 4 selected sboms. e.g. `PUT /user-preferences/selected-sboms` with a body like
> 
> ```json
> [
> {
> "sbom_id": "abcd",
> "sbom_name": "xyz"
> },
> {
> "sbom_id": "abcd",
> "sbom_name": "xyz"
> },
> {
> "sbom_id": "abcd",
> "sbom_name": "xyz"
> },
> {
> "sbom_id": "abcd",
> "sbom_name": "xyz"
> }
> ]
> ```
> 
> The token always arrives to the backend and the backend could decode the token and extract the username or userid, or any other field that we might find appropriate.
> 
> It would be nice to have the whole thing done in this PR so we avoid having multiple PRs.
> 
> What do you think?

I added  endpoint and helm change, helm needs test.
I think we do not needs add special one like selected-sboms, The user_preferences is a whole, and we only need to access the entire user_preferences by using the ID.

**Review by carlosthe19916 — CHANGES_REQUESTED** [review]
@bxf12315 Thanks for the PR. Here are some things I suggest changing please:

### How the UI expects calls to be done.

- The flow/order of calls for the UI will be:
  - Step1: get the current user preferences `GET /api/v1/dashboard/userPreferences/{USERNAME}`. Currently if I do it in an fresh instance of trustify I get the error: `500 Internal server error: no rows returned by a query that expected to return at least one row` 
  - Step2: use the data that came from the backend and render it appropriately in the UI
  - Step3: If the user wants he can update his preferences.

### Here are the things I suggest changing

- We already have this endpoint: `GET /api/v1/dashboard/userPreferences/{USERNAME}`:
  - Regardless of whether or not the user have some data saved, the endpoint should ALWAYS return the following JSON:
```json
{
  "sbom1": string | undefined,
  "sbom2": string | undefined,
  "sbom3": string | undefined,
  "sbom4": string | undefined
}
```

- In regards of how to update the user preferences we can do `POST /api/v1/dashboard/userPreferences/{USERNAME}` and the UI will always send an sbom equal to the previous GET endpoint:

```json
{
  "sbom1":  "my_sbom_id_1",
  "sbom2":  "my_sbom_id_2",
  "sbom3":  "my_sbom_id_3"
  "sbom4":  "my_sbom_id_4"
}
```

So it would be good not to accept "any" string from the UI but only accept a json with the schema json that I shared above. Then, that json can be stored in the db as plain string, but we still need a defined rust struct for the user preferences (the one I shared above)

**Comment by carlosthe19916** [comment]
Or rather than:
```json
{
  "sbom1":  "my_sbom_id_1",
  "sbom2":  "my_sbom_id_2",
  "sbom3":  "my_sbom_id_3"
  "sbom4":  "my_sbom_id_4"
}
```

we could use:
```json
{
  "sbom1_id":  "my_sbom_id_1",
  "sbom2_id":  "my_sbom_id_2",
  "sbom3_id":  "my_sbom_id_3"
  "sbom4_id":  "my_sbom_id_4"
}
```

File: spog/api/src/endpoints/dashboard.rs [hunk 1]
```diff
@@ -1,26 +1,75 @@
...
+pub(crate) fn configure(auth: Option<Arc<Authenticator>>, payload_limit: usize) -> impl FnOnce(&mut ServiceConfig) {
+    move |config: &mut ServiceConfig| {
         config.service(
             web::scope("/api/v1/dashboard")
                 .wrap(new_auth!(auth))
-                .service(web::resource("/status").to(get_status)),
+                .service(web::resource("/status").to(get_status))
+                .service(
+                    // web::scope("/api/v1/dashboard")
+                    web::resource("/userPreferences/{id}").to(user_preferences_receive),
```
Comment (mrizzi) [inline spog/api/src/endpoints/dashboard.rs:22]: Sorry guys for jumping into the discussion directly but I've read about the `/api/v1/dashboard/userPreferences/{USERNAME}` endpoint.
If I'm not misunderstanding it or missing something, I strongly suggest you to avoid having the username as a parameter but instead use the username actually logged to retrieve his/her preferences (if possible).
Basically an authenticated call to the `/api/v1/dashboard/userPreferences/` endpoint should be enough to retrieve the user doing the authenticated call itself.
Otherwise I'm concerned about having to guarantee no information can be leaked between users.

**Review by mrizzi — CHANGES_REQUESTED** [review]
Added a consideration related to authentication and authorization managed in the user preferences GET endpoint that I believe it's better to clarify before moving forwards.

File: spog/api/src/endpoints/dashboard.rs (hunk 1 above)
Comment (ctron) [reply spog/api/src/endpoints/dashboard.rs:22]: I think the same should be true for the other endpoints.

File: deploy/k8s/charts/trustification/templates/services/spog/api/030-Deployment.yaml [hunk 2]
```diff
@@ -32,12 +32,14 @@ spec:
...
 
-          command: ["/trust"]
+          command: [ "/trust" ]
           args:
             - "spog"
             - "api"
             - "-p"
             - "8080"
+            - "--db-storage-base"
+            - "{{ include "trustification.preferences.db.path" $mod}}"
```
Comment (ctron) [inline deploy/k8s/charts/trustification/templates/services/spog/api/030-Deployment.yaml:42]: What will happen when the user decides to scale up this deployment?

Comment (mrizzi) [reply deploy/k8s/charts/trustification/templates/services/spog/api/030-Deployment.yaml:42]: That's a good point.

With this approach based on SQLite it has been accepted, like it happens for the collectorist component, the "compromise" of losing the ability to scale up these components based on the considerations that:

- the performance tests gave good enough results with the current default configuration
- there's no plan, considered the trustification's lifespan, to invest further efforts in testing the scaling up of the current architecture components

It's very opinionated, we all know, but I hope at least this makes sense.

Comment (ctron) [reply deploy/k8s/charts/trustification/templates/services/spog/api/030-Deployment.yaml:42]: We should just be clear about that in the documentation.

**Comment by carlosthe19916** [comment]
@bxf12315 may I ask you please to synchronize this branch with the latest changes of the main branch?

**Comment by bxf12315** [comment]
> @bxf12315 may I ask you please to synchronize this branch with the latest changes of the main branch?



> @bxf12315 may I ask you please to synchronize this branch with the latest changes of the main branch?

Done

**Comment by carlosthe19916** [comment]
@mrizzi After my previous comment are addressed I was planning to squash and merge this PR. Is there anything you would recommend in regards to the changes to the helm chart done in this PR? I can see the gh tests of helm is failing and acknowledge the fact that you worked on the helm charts more than me so your opinion would be very welcome

File: deploy/k8s/charts/trustification-infrastructure/Chart.yaml [hunk 3]
```diff
@@ -20,7 +20,7 @@ home: https://trustification.io
 dependencies:
   - name: keycloak
     repository: https://charts.bitnami.com/bitnami
-    version: 18.2.1
+    version: 22.2.1
```
Comment (mrizzi) [inline deploy/k8s/charts/trustification-infrastructure/Chart.yaml:23]: I feel like this update is out of scope for this PR.
If confirmed, please remove it.

**Review by mrizzi — CHANGES_REQUESTED** [review]
The helm chart test must work.
I've spotted an unexpected change in one chart dependency.
Please remove it if not needed and then, if needed, apply further changes to have the helm chart working fine.

File: deploy/k8s/charts/trustification-infrastructure/Chart.yaml (hunk 3 above)
Comment (bxf12315) [reply deploy/k8s/charts/trustification-infrastructure/Chart.yaml]: yes I will change it.

**Review by carlosthe19916 — APPROVED** [review]
@mrizzi @bxf12315 this PR looks good to me! Thanks @bxf12315 !
@mrizzi the helm tests in gh actions are now passing. If there are no objections we could merge this PR

**Review by mrizzi — APPROVED** [review]
@bxf12315 thanks, the helm chart looks better.
@carlosthe19916 please go ahead and merge it.
